message.go

```diff
@@ -54,203 +66,350 @@ type Emote struct {
+	if len(split) < 2 {
+		return &message{
+			RawMessage: rawMessage{
+				Type:    ERROR,
+				Raw:     line,
+				Message: line,
+			},
+		}
 	}
-	msg.Raw = line
-	return msg
-}
 
-func parseOtherMessage(line string) *message {
-	msg := &message{
-		Type: UNSET,
+	// If there is only two values, message is empty. Fill it with a blank string so we can assign it to rawMessage.
+	if len(split) == 2 {
+		for i := 0; i < 3-len(split); i++ {
+			split = append(split, "")
+		}
 	}
-	split := strings.Split(line, " ")
-	msg.Raw = line
 
-	msg.Type = parseMessageType(split[2])
-	msg.Tags = make(map[string]string)
+	rawType, channel, username := parseMiddle(split[1])
+
+	rawMessage := rawMessage{
+		Type:    parseMessageType(rawType),
+		RawType: rawType,
+		Raw:     line,
+		Tags:    parseTags(split[0]),
+		Message: split[2],
+	}
 
-	// Parse out channel if it exists in this line
-	if len(split) >= 4 && len(split[3]) > 1 && split[3][0] == '#' {
-		// Remove # from channel
-		msg.Channel = split[3][1:]
+	return &message{
+		Channel:    channel,
+		Username:   username,
+		RawMessage: rawMessage,
 	}
+}
 
-	tagsString := strings.Fields(strings.TrimPrefix(split[0], "@"))
-	tags := strings.Split(tagsString[0], ";")
-	for _, tag := range tags {
-		tagSplit := strings.Split(tag, "=")
+func parseMiddle(middle string) (string, string, string) {
+	var rawType, channel, username string
 
-		value := ""
-		if len(tagSplit) > 1 {
-			value = tagSplit[1]
+	for i, v := range strings.SplitN(middle, " ", 3) {
+		switch {
+		case i == 1:
+			rawType = v
+		case strings.Contains(v, "!"):
+			username = strings.SplitN(v, "!", 2)[0]
+		case strings.Contains(v, "#"):
+			channel = strings.TrimPrefix(v, "#")
 		}
-
-		msg.Tags[tagSplit[0]] = value
 	}
 
-	if msg.Type == CLEARCHAT {
-		msg.Text = "Chat has been cleared by a moderator"
-	}
-	return msg
+	return rawType, channel, username
 }
 
 func parseMessageType(messageType string) MessageType {
 	switch messageType {
-	case "PRIVMSG":
-		return PRIVMSG
 	case "WHISPER":
 		return WHISPER
+	case "PRIVMSG":
+		return PRIVMSG
 	case "CLEARCHAT":
 		return CLEARCHAT
-	case "NOTICE":
-		return NOTICE
 	case "ROOMSTATE":
 		return ROOMSTATE
-	case "USERSTATE":
-		return USERSTATE
 	case "USERNOTICE":
 		return USERNOTICE
+	case "USERSTATE":
+		return USERSTATE
+	case "NOTICE":
+		return NOTICE
 	default:
 		return UNSET
 	}
 }
 
-func parseMiddle(middle string) (string, MessageType, string) {
-	var username string
-	var msgType MessageType
-	var channel string
+func parseTags(tagsRaw string) map[string]string {
+	tags := make(map[string]string)
 
-	for i, c := range middle {
-		if c == '!' {
-			username = middle[:i]
-			middle = middle[i:]
+	tagsRaw = strings.TrimPrefix(tagsRaw, "@")
+	for _, v := range strings.Split(tagsRaw, ";") {
+		tag := strings.SplitN(v, "=", 2)
+
+		var value string
+		if len(tag) > 1 {
+			value = tag[1]
```

Comment:
This value should be unescaped first: https://ircv3.net/specs/extensions/message-tags.html#escaping-values

Reply:
Just to comment on why i've specifically picked this: There are users on twitch with spaces in their display names, and that value would stay escapes as `\s` if we don't unescape the value first.